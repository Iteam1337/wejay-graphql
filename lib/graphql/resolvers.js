import SpotifyWebApi from 'spotify-web-api-node'

require('dotenv').config()

const rooms = [
  {
    id: '1',
    name: 'Test',
    queue: [],
    users: []
  },
  {
    id: '2',
    name: 'Iteam',
    queue: [],
    users: []
  }
]

const spotifyApi = new SpotifyWebApi({
  clientId: process.env.CLIENT_ID,
  clientSecret: process.env.CLIENT_SECRET
})

spotifyApi.clientCredentialsGrant().then(
  data => {
    console.log('The access token expires in ' + data.body['expires_in'])
    console.log('The access token is ' + data.body['access_token'])

    // Save the access token so that it's used in future calls
    spotifyApi.setAccessToken(data.body['access_token'])
  },
  err => {
    console.log(
      'Something went wrong when retrieving an access token',
      err.message
    )
  }
)

export const resolvers = {
  /// Queries
  Query: {
    rooms: () => {
      return rooms
    },
    room: (_, { id }) => {
      return rooms.find(room => room.id === id)
    }
  },

  /// Mutations
  Mutation: {
    queueTrack: async (_, { input }) => {
      const { roomId } = input
      let { spotifyId } = input

      if (spotifyId.includes('spotify:track')) {
        const parsedId = spotifyId.split(':')
        spotifyId = parsedId[parsedId.length - 1]
      }

      const room = rooms.find(room => room.id === roomId)

      if (!room) {
        throw new Error('Room does not exist')
      }

      const { body } = await spotifyApi.getTrack(spotifyId)

      const newTrack = {
        album: body.album,
        artists: body.artists,
        duration: body.duration_ms,
        name: body.name,
        spotifyUri: body.uri
      }

      room.queue.push(newTrack)

      return newTrack
    }
  }
}
