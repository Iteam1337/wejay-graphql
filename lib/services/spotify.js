import SpotifyWebApi from 'spotify-web-api-node'

require('dotenv').config()

const spotifyApi = new SpotifyWebApi({
  clientId: process.env.CLIENT_ID,
  clientSecret: process.env.CLIENT_SECRET
})

let timer

spotifyApi.clientCredentialsGrant().then(
  data => {
    if (data.body) {
      console.log('updated token - ', Date.now())

      spotifyApi.setAccessToken(data.body['access_token'])

      timer = setInterval(() => {
        spotifyApi.clientCredentialsGrant().then(data => {
          console.log('updated token - ', Date.now())

          spotifyApi.setAccessToken(data.body['access_token'])
        })
      }, data.body['expires_in'] * 1000)
    }
  },
  err => {
    console.log(
      'Something went wrong when retrieving an access token',
      err.message
    )
  }
)

export default spotifyApi
