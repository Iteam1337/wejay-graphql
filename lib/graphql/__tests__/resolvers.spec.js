import MockDate from 'mockdate'

import spotify from '../../services/spotify'
import { resolvers } from '../resolvers'

jest.mock('../../services/spotify', () => ({
  getTrack: jest.fn()
}))

jest.mock('uuid/v4', () => jest.fn(() => '1'))

describe('resolvers', () => {
  describe('Query', () => {
    describe('#rooms', () => {
      it('returns current rooms', () => {
        expect(resolvers.Query.rooms()).toMatchSnapshot()
      })
    })

    describe('#room', () => {
      it('throws an error if no room is found', async () => {
        try {
          await resolvers.Query.room(null, { roomName: 'testz' })
        } catch (e) {
          expect(e.message).toMatch(/Room does not exist/)
        }
      })

      it('finds one room using id', () => {
        resolvers.Mutation.addRoom(null, { roomName: 'test' })

        expect(resolvers.Query.room(null, { name: 'test' })).toMatchSnapshot()
      })
    })
  })

  describe('Mutations', () => {
    describe('#queueTrack', () => {
      it('throws an error if no room is found', async () => {
        try {
          const input = {
            roomName: 'testz',
            spotifyId: 'spotify:track:6Inmo6ElDCaiZCiB6lficH'
          }

          await resolvers.Mutation.queueTrack(null, { input })
        } catch (e) {
          expect(e.message).toMatch(/Room does not exist/)
        }
      })

      it('gets the track information from the Spotify API', async () => {
        try {
          const input = {
            roomName: 'test',
            spotifyId: 'spotify:track:6Inmo6ElDCaiZCiB6lficH'
          }

          spotify.getTrack.mockImplementationOnce(() =>
            Promise.resolve({ body: {} })
          )

          resolvers.Mutation.addRoom(null, { roomName: 'test' })

          await resolvers.Mutation.queueTrack(null, { input })

          expect(spotify.getTrack).toHaveBeenCalledWith(
            '6Inmo6ElDCaiZCiB6lficH'
          )
        } catch (e) {
          throw new Error(e)
        }
      })

      it('returns the new track', async () => {
        try {
          const input = {
            roomName: 'test',
            spotifyId: 'spotify:track:6Inmo6ElDCaiZCiB6lficH'
          }

          spotify.getTrack.mockImplementationOnce(() =>
            Promise.resolve({
              body: {
                album: {
                  name: 'albumname'
                },
                artists: [{ name: 'artist' }],
                duration_ms: 30000,
                name: 'track',
                uri: '6Inmo6ElDCaiZCiB6lficH'
              }
            })
          )

          resolvers.Mutation.addRoom(null, { roomName: 'test' })

          const queuedTrack = await resolvers.Mutation.queueTrack(null, {
            input
          })

          expect(queuedTrack).toMatchSnapshot()
        } catch (e) {
          throw new Error(e)
        }
      })
    })

    describe('#addRoom', () => {
      it('returns a new room', () => {
        expect(
          resolvers.Mutation.addRoom(null, { roomName: 'newRoom' })
        ).toMatchSnapshot()
      })
    })

    describe('#nextTrack', () => {
      it('throws an error if no room is found', async () => {
        try {
          const input = {
            roomName: 'testz'
          }

          await resolvers.Mutation.nextTrack(null, input)
        } catch (e) {
          expect(e.message).toMatch(/Room does not exist/)
        }
      })
    })

    describe('#joinRoom', () => {
      afterEach(() => {
        MockDate.reset()
      })

      it('throws an error if no room is found', async () => {
        try {
          const input = {
            roomName: 'test',
            email: 'cookie@monster.com'
          }

          await resolvers.Mutation.joinRoom(null, { input })
        } catch (e) {
          expect(e.message).toMatch(/Room does not exist/)
        }
      })

      it('returns the updated room', async () => {
        MockDate.set('2017-09-20')

        try {
          const input = {
            roomName: 'test',
            email: 'cookie@monster.com'
          }

          resolvers.Mutation.addRoom(null, { roomName: 'test' })

          const updatedRoom = await resolvers.Mutation.joinRoom(null, { input })

          expect(updatedRoom).toMatchSnapshot()
        } catch (e) {
          throw new Error(e)
        }
      })
    })
  })
})
