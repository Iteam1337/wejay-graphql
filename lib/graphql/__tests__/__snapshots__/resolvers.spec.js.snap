// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvers Mutations #addRoom returns a new room 1`] = `
Room {
  "currentTrack": null,
  "history": Array [],
  "id": "1",
  "name": "newRoom",
  "queue": Array [],
  "timer": null,
  "users": Array [],
}
`;

exports[`resolvers Mutations #joinRoom returns the updated room 1`] = `
Room {
  "currentTrack": null,
  "history": Array [],
  "id": "1",
  "name": "test",
  "queue": Array [
    Object {
      "album": undefined,
      "artists": undefined,
      "duration": undefined,
      "name": undefined,
      "spotifyUri": undefined,
    },
    Object {
      "album": Object {
        "name": "albumname",
      },
      "artists": Array [
        Object {
          "name": "artist",
        },
      ],
      "duration": 30000,
      "name": "track",
      "spotifyUri": "6Inmo6ElDCaiZCiB6lficH",
    },
  ],
  "timer": null,
  "users": Array [
    Object {
      "email": "cookie@monster.com",
    },
    Object {
      "email": "cookie@monster.com",
    },
  ],
}
`;

exports[`resolvers Mutations #queueTrack returns the new track 1`] = `
Object {
  "album": Object {
    "name": "albumname",
  },
  "artists": Array [
    Object {
      "name": "artist",
    },
  ],
  "duration": 30000,
  "name": "track",
  "spotifyUri": "6Inmo6ElDCaiZCiB6lficH",
}
`;

exports[`resolvers Query #room finds one room using id 1`] = `
Room {
  "currentTrack": null,
  "history": Array [],
  "id": "1",
  "name": "test",
  "queue": Array [],
  "timer": null,
  "users": Array [],
}
`;

exports[`resolvers Query #rooms returns current rooms 1`] = `Array []`;
